### PR TITLE
Add pipe_warning() and deprecate glue_pipe arg

### LINKED ISSUE #184: Add glue_pipe argument to grade_this_code()
Bring support for the `glue_pipe` warning to `grade_this_code()`.

https://github.com/rstudio-education/gradethis/blob/b2d9ef55d3f68d3dcadae06df2fc213d87f2d30b/R/grade_this_code.R#L41-L43

We should also ensure that the message displays the _unpiped_ code, rather than the piped user code.

```r
exercise <- list2env(list(
  .user_code = "penguins %>% pull(year) %>% min(year)",
  .solution_code = "penguins %>% pull(year) %>% min()"
))

grade_code()(exercise)
#> <gradethis_graded: Incorrect: I see that you are using pipe operators (e.g. %>%), so I want to let you know that this is how I am interpretting your code before I check it:
#> 
#> "penguins %>% pull(year) %>% min(year)"
#> 
#> I did not expect your call to min() to include year. You may have included an unnecessary argument, or you may have left out or misspelled an important argument name.  Let's try it again.>

grade_this_code()(exercise)
#> <gradethis_graded: Incorrect: I did not expect your call to min() to include year. You may have included an unnecessary argument, or you may have left out or misspelled an important argument name. Try it again. Perseverence is the key to success.>
```

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #184

The pipe warning message is now glued with access only to `.user_code` and `.user_code_unpiped` (see below). It then becomes an additional magic glue variable available for use in failure messages in `grade_code()` and `grade_this_code()` as `.message_pipe_warning`.

I think this will make it easier for users to compose glue template strings for messages and it also decouples `grade_code()` and `grade_this_code()` to smooth out the subtle inconsistencies in how they compose the `.message`.

So this glue warning becomes `.message_pipe_warning` if one is needed

```
gradethis_glue_pipe = paste0(
  "I see that you are using pipe operators (e.g. %>%), ",
  "so I want to let you know that this is how I am interpretting your code ",
  "before I check it:\n\n```r\n{.user_code_unpiped}\n```\n\n"
)
```

for use in the incorrect messages templates as follows:

```r
# Default message for grade_this_code(incorrect)
gradethis.code.incorrect = "{.message_pipe_warning}{.message} {random_encouragement()}"

# Default for incorrect legacy grade_code(), grade_result() and grade_result_strict()
gradethis_glue_incorrect = "{.message_pipe_warning}{.message} {.incorrect} {random_encouragement()}"
```

### Overview of changes

1. Created internal `glue_pipe_message` helper function for adding pipe message
    - Setting the `gradethis_glue_pipe` to `NULL` or `""` disables the pipe message
2. Standardize the magic glue variables for this message:
    - User code as written is now in `.user_code` rather than `.user`
    - Unpiped user code is available in `.user_code_unpiped` which avoids users having to rely on undocumented and non-exported functions (`unpipe_all_str()`) in the glue message
3. `grade_this_code()` always returns code feedback, so we don't need the `fail_code_feedback` argument
    - I updated the tests to test `maybe_code_feedback()` in a `grade_this()` context where it's intended to be used.
4. `grade_result()` and `grade_result_strict()` set `.message_pipe_warning = NULL` since they both use `gradethis_glue_incorrect`.

### Questions

Should we update the global option from the old style `gradethis_glue_pipe` to `gradethis.glue_pipe` or `gradethis.code.glue_pipe`?

I also started to wonder how much of this is necessary if we're showing the submitted piped code in `code_feedback()` with #192

### CODE REVIEW

**Comment by schloerke** [comment]
> Should we update the global option from the old style `gradethis_glue_pipe` to `gradethis.glue_pipe` or `gradethis.code.glue_pipe`?

How about `gradethis.code.pipe`? As the others do not have `glue` in the name, such as `gradethis.code.correct`.

> I also started to wonder how much of this is necessary if we're showing the submitted piped code in code_feedback() with #192

That is very fair. Maybe we do the change now, but we can remove the `.message_pipe_warning` later.

**Review by schloerke — CHANGES_REQUESTED** [review]
.

**Comment by gadenbuie** [comment]
> How about `gradethis.code.pipe`? As the others do not have `glue` in the name, such as `gradethis.code.correct`.

I'd really like the suffix to match the argument name when one exists. E.g. `gradethis.code.correct` matches `correct` parameter of `grade_this_code()`.

This brings up that the new `grade_this*()` don't use the `glue_` prefix, so I'm wondering if we should rename the parameter `pipe_warning` in `grade_this_code()`. This would create symmetry between `.message_pipe_warning` and the `pipe_warning` argument. As we push `grade_result()` and `grade_code()` to the back burner, the parameter names don't need to be the same.

OTOH, the whole point of the pipe warning was to help make sense of the code feedback, but I'm starting to think that we're close enough to the code feedback messages that we want that maybe we could just take out the pipe warning. I'd have to audit the scenarios where the pipe message appears first though.

**Comment by schloerke** [comment]
So rather than adding more parameters, what about adding it into the incorrect message?

I agree that there is some _magic variables_ with `.message` and now `.message_pipe_warning`.

What about updating to
```r
gradethis.code.incorrect = "{pipe_warning()}{code_feedback()} {random_encouragement()}"
```

The only reason for `.message` to exist is due to `allow_partial_matching` being fed into `code_feedback()`. At this point, I'd rather have `code_feedback()` recalculated in the message as it is fast enough.


---------------

> I'd like the suffix to match the argument name

How about
```r
partial_matching = getOption("gradethis.code.partial_matching", TRUE)
```

---------------------

I'd be really happy if we could make it not be required to repeat a variable for `code_feedback()` when called inside `grade_this_code()`.


Don't know if can update the default value to : 
```
code_feedback(
  partial_matching = get0(".__partial_matching", parent.frame(), ifnotfound = getOption("gradethis.code.partial_matching", TRUE))
)
```
.  This would allow for `gradethis.code.incorrect` to default to `"{pipe_warning()}{code_feedback()} {random_encouragement()}"` as all the information would be provided in the calling environment

**Comment by gadenbuie** [comment]
@schloerke I really like the idea of making `pipe_warning()` and `code_feedback()` functions that are used in glue-able messages. I think it makes the messages easier to reason about. What do you think of the approach in [87a859f](https://github.com/rstudio-education/gradethis/pull/193/commits/87a859fbc183127c9f66b3c6c7f646df1b1a169b)?

- Change message option to `gradethis.pipe_warning()`
- Add `pipe_warning()` to default message options
- Deprecate `glue_pipe` argument of `glue_code()`

If that looks good to you, I'll fill in the documentation of `pipe_warning()` and move on getting `code_feedback()` into the same place.

**Comment by gadenbuie** [comment]
@schloerke I updated `gradethis.code.incorrect` to default to 

```
gradethis.code.incorrect = "{pipe_warning()}{code_feedback()} {random_encouragement()}"
```

using `with_options()` to ensure that `allow_partial_matching` is passed from `grade_this_code()` to `code_feedback()`.

**Comment by gadenbuie** [comment]
I really like this approach. I'm okay with leaving a few variables that are set in the glue environment in the legacy code, but I like the design philosophy of using documented, exported functions in the glue strings. I think that will ultimately be more transparent to end users.

File: R/zzz.R [hunk 1]
```diff
@@ -18,17 +18,17 @@ gradethis_default_options <- list(
...
+  gradethis_glue_correct = "{random_praise()} {.message} {.correct}",
+  gradethis_glue_incorrect = "{pipe_warning()}{.message} {.incorrect} {random_encouragement()}",
 
-  gradethis_glue_pipe = paste0(
+  gradethis.pipe_warning = paste0(
     "I see that you are using pipe operators (e.g. %>%), ",
     "so I want to let you know that this is how I am interpretting your code ",
-    "before I check it:\n\n```r\n{unpipe_all_str(.user, width = 60)}\n```\n\n{.message}"
+    "before I check it:\n\n```r\n{.user_code_unpiped}\n```\n\n"
   ),
```
Comment (schloerke) [inline R/zzz.R:28]: Now that this is not _legacy_, should we move it up the the upper group?

File: R/zzz.R [hunk 2]
```diff
@@ -18,17 +18,17 @@ gradethis_default_options <- list(
 
 
   ### legacy ###
-  gradethis_glue_correct = "{ random_praise() } { .message } { .correct }",
-  gradethis_glue_incorrect = "{ .message } { .incorrect } { random_encouragement() }",
+  gradethis_glue_correct = "{random_praise()} {.message} {.correct}",
+  gradethis_glue_incorrect = "{pipe_warning()}{.message} {.incorrect} {random_encouragement()}",
 
-  gradethis_glue_pipe = paste0(
```
Comment (schloerke) [inline R/zzz.R:24]: Should we put in the news that this is removed?

**Review by schloerke — APPROVED** [review]

File: R/zzz.R (hunk 1 above)
Comment (gadenbuie) [reply R/zzz.R:28]: This one would be the exception to that rule. Technically, `pipe_warning()` creates the message but it assembles `.user_code_unpiped` inside the function. I don't think we want to export `unpipe_all_str()`, so I think we're stuck with it.

I guess I'd amend the rule to say that any helpers used in the default arguments of top-level functions should also be functions. So in this case `pipe_warning()` is called at the top level in the glue-string of `gradethis.code.incorrect`, but `gradethis.pipe_warning` is the default value of `message` in `pipe_warning()`, or two levels deep, so it's okay if it uses a non-function value in the glue string.

File: R/zzz.R (hunk 2 above)
Comment (gadenbuie) [reply R/zzz.R:24]: Good idea, thanks!

File: R/zzz.R (hunk 1 above)
Comment (schloerke) [reply R/zzz.R:28]: My fault for not being clear.

Should we move the `gradethis.pipe_warning` definition above line 20 as `gradethis.pipe_warning` is not just for soft deprecated functions?

Comment (gadenbuie) [reply R/zzz.R]: oh I see, yeah that makes sense

File: R/grade_code.R [hunk 3]
```diff
@@ -83,13 +83,24 @@ grade_code <- function(
...
+  glue_pipe = deprecated(),
   grader_args = deprecated(),
   learnr_args = deprecated()
 ) {
   ellipsis::check_dots_empty()
   if (is_present(grader_args)) deprecate_warn("0.2.0", "grade_code(grader_args = )")
   if (is_present(learnr_args)) deprecate_warn("0.2.0", "grade_code(learnr_args = )")
+  if (is_present(glue_pipe)) {
+    deprecate_warn(
+      "0.2.1",
```
Comment (gadenbuie) [inline R/grade_code.R:95]: @schloerke Is this value okay? If so I'd bump gradethis to `0.2.1.9000` before merging.

Comment (schloerke) [reply R/grade_code.R:95]: Sounds good

**Pull request merged**